maven-plugin/src/main/java/io/confluent/kafka/schemaregistry/maven/derive/schema/DeriveProtobufSchema.java

```diff
@@ -0,0 +1,143 @@
+    classToDataType.put(com.fasterxml.jackson.databind.node.TextNode.class.getName(), STRING);
+    classToDataType.put(com.fasterxml.jackson.databind.node.BigIntegerNode.class.getName(), DOUBLE);
+    classToDataType.put(com.fasterxml.jackson.databind.node.IntNode.class.getName(), INT_32);
+    classToDataType.put(com.fasterxml.jackson.databind.node.LongNode.class.getName(), INT_64);
+    classToDataType.put(com.fasterxml.jackson.databind.node.BooleanNode.class.getName(), BOOL);
+    classToDataType.put(com.fasterxml.jackson.databind.node.NullNode.class.getName(), ANY_FIELD);
+    classToDataType.put(com.fasterxml.jackson.databind.node.MissingNode.class.getName(), ANY_FIELD);
+  }
+
+  protected ObjectNode mergeMultipleDataTypes(ObjectNode mergedArray,
+                                              List<JsonNode> primitives,
+                                              List<JsonNode> records,
+                                              List<JsonNode> arrays,
+                                              boolean check2dArray) {
+    ArrayNode items = mapper.createArrayNode();
+    // Check for nested array, if yes return error
+    if (arrays.size() > 0 && check2dArray) {
+      throw new IllegalArgumentException(String.format("found nested array: %s", arrays));
+    }
+
+    DeriveSchemaUtils.mergeNumberTypes(primitives);
+    // Adding primitive types to items' list
+    items.addAll(DeriveSchemaUtils.getUnique(primitives));
+    // No merging of arrays, add directly to items' list
+    items.addAll(arrays);
+
+    // Merge records if there is at least 1 record
+    if (records.size() > 0) {
+      items.add(mergeRecords(records));
+    }
+
+    if (items.size() > 1) {
+      throw new IllegalArgumentException(String.format("found multiple data types: %s", items));
+    } else if (items.size() == 1) {
+      mergedArray.set("items", items.get(0));
+    } else {
+      mergedArray.set("items", getNullSchema());
+    }
+    return mergedArray;
+  }
+
+  @Override
+  public TextNode convertToFormat(JsonNode schema, String name) {
+    // Json schema is used as a template. Data types and array merging is defined specifically for
+    // protobuf(mergeMultipleDataTypes and classToDataType), this converts json template to protobuf
+    String schemaForRecord = convertToFormatRecord(schema, name);
+    StringBuilder schemaBuilder = new StringBuilder(PROTOBUF_SYNTAX);
+    if (schemaForRecord.contains(ANY_FIELD)) {
+      schemaBuilder.append("import \"google/protobuf/any.proto\";\n");
+    }
+    schemaBuilder.append(schemaForRecord);
+    ProtobufSchema protobufSchema = new ProtobufSchema(schemaBuilder.toString());
+    protobufSchema.validate();
+    return mapper.convertValue(protobufSchema.toString(), TextNode.class);
+  }
+
+  protected String convertToFormatRecord(JsonNode schema, String name) {
+    // Convert json schema for record to protobuf schema
```

Comment:
A general comment to add the description of the function (what it does) outside of the function on top of it instead of just inside it like here.

Reply:
Should I make this change for all the functions then?

Reply:
yes for this PR atleast we should cover the files it touches.

Reply:
Sure!